src/Behat/Behat/HelperContainer/Argument/ServicesResolverFactory.php

```diff
@@ -0,0 +1,169 @@
+<?php
+
+/*
+ * This file is part of the Behat.
+ * (c) Konstantin Kudryashov <ever.zet@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Behat\Behat\HelperContainer\Argument;
+
+use Behat\Behat\Context\Argument\SuiteScopedResolverFactory;
+use Behat\Behat\HelperContainer\BuiltInServiceContainer;
+use Behat\Behat\HelperContainer\Exception\WrongContainerClassException;
+use Behat\Behat\HelperContainer\Exception\WrongServicesConfigurationException;
+use Behat\Behat\HelperContainer\ServiceContainer\HelperContainerExtension;
+use Behat\Testwork\Suite\Suite;
+use Interop\Container\ContainerInterface;
+use Symfony\Component\DependencyInjection\TaggedContainerInterface;
+
+/**
+ * Generates ServiceContainer argument resolvers based on suite's `services` setting.
+ *
+ * @see ContextEnvironmentHandler
+ *
+ * @author Konstantin Kudryashov <ever.zet@gmail.com>
+ */
+final class ServicesResolverFactory implements SuiteScopedResolverFactory
+{
+    /**
+     * @var TaggedContainerInterface
+     */
+    private $container;
+
+    /**
+     * Initialises factory.
+     *
+     * @param TaggedContainerInterface $container
+     */
+    public function __construct(TaggedContainerInterface $container)
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function generateArgumentResolvers(Suite $suite)
+    {
+        if (!$suite->hasSetting('services')) {
+            return array();
+        }
+
+        $container = $this->createContainer($suite->getSetting('services'));
+
+        return array($this->createArgumentResolver($container));
+    }
+
+    /**
+     * Creates container from the setting passed.
+     *
+     * @param string $settings
+     *
+     * @return mixed
+     */
+    private function createContainer($settings)
+    {
+        if (is_string($settings)) {
+            return $this->createContainerFromString($settings);
+        }
+
+        if (is_array($settings)) {
+            return $this->createContainerFromArray($settings);
+        }
+
+        throw new WrongServicesConfigurationException(
+            sprintf('`services` must be either string or an array, but `%s` given.', gettype($settings))
+        );
+    }
+
+    /**
+     * Creates custom container using class/constructor given.
+     *
+     * @param string $settings
+     *
+     * @return mixed
+     */
+    private function createContainerFromString($settings)
+    {
+        if ('@' === mb_substr($settings, 0, 1)) {
```

Comment:
A similar check takes place in `ServicesResolverFactory` - can these be done in one place? Perhaps a value object?